### PR TITLE
gh-109370: Support closing Connection and PipeConnection from other thread

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-109370

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
I do not like this.

1. It is too complicated and verbose.
2. There is still a chance for race condition, if between pushing `self.handler` on the stack and calling the corresponding OS function, the handler was closed and a new file was open that reuses the same fileno.

I have not tested the code on Windows yet, very likely more try/exept or return code checks should be added in Windows specific code.

File: Lib/multiprocessing/connection.py [hunk 1]
```diff
@@ -341,20 +356,44 @@ def _recv_bytes(self, maxsize=None):
...
         def _poll(self, timeout):
-            if (self._got_empty_message or
-                        _winapi.PeekNamedPipe(self._handle)[0] != 0):
-                return True
+            try:
+                if (self._got_empty_message or
+                            _winapi.PeekNamedPipe(self._handle)[0] != 0):
+                    return True
+            except OSError as err:
+                if err.errno == errno.EBADF:
```
Comment (vstinner) [inline Lib/multiprocessing/connection.py:364]: We should not get EBADF. We should just *not* call functions on closed handle. If we land in this case, we failed badly before, no?

File: Lib/multiprocessing/connection.py [hunk 2]
```diff
@@ -275,19 +274,27 @@ class PipeConnection(_ConnectionBase):
...
-            ov, err = _winapi.WriteFile(self._handle, buf, overlapped=True)
+            try:
+                ov, err = _winapi.WriteFile(self._handle, buf, overlapped=True)
+            except OSError as err:
+                if err.errno == errno.EBADF:
+                    raise OSError(errno.EPIPE, "handle is closed")
+                raise
+            except TypeError:
+                self._check_closed()
+                raise
```
Comment (vstinner) [inline Lib/multiprocessing/connection.py:297]: Would it be possible to split this change in two parts? Keep this change for _check_closed() change which now raises BrokenPipeError, and you changes around close(): these ones look straightforward are correct. But write a separated PR to add these try/except?

I dislike these try/except. If there is a risk to land into EBADF case with the current code, maybe some kind of locking is needed to *not* call the Windows API with a handle, while another thread call close() which invalidates the handle.

File: Lib/multiprocessing/connection.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/multiprocessing/connection.py:364]: How can we know that it was closed in other thread without using locks?

You can get EBADF in the following scenario:

* Thread A reads `self._handle` and puts in on the stack.
* Thread B closes it.
* Thread A calls function with an already closed file descriptor on the stack.
* That function fails with EBADF errno.

It is the bast case. The worst case is:

* Thread A reads `self._handle` and puts in on the stack.
* Thread B closes it.
* Thread B or C opens a new file descriptor which reuses the same integer value.
* Thread A calls function with a file descriptor  which now refers to unrelated file or socket.
* Unpredictable consequences.

**Comment by serhiy-storchaka** [comment]
The more I look at this, the more I dislike it. I do not see how is it possible to safely close a file descriptor on other thread without locking. Making Connection and PipeConnection truly thread-safe in a task on completely different level.

I opened #109780 instead.

**Comment by vstinner** [comment]
> Making Connection and PipeConnection truly thread-safe in a task on completely different level.

Maybe we should just give up to document that they are not thread-safe.

I recall that I did that once on some asyncio classes.

**Pull request closed without merging**